app/controllers/admin/carbon_sources_controller.rb

```diff
@@ -2,29 +2,62 @@ module Admin
   class CarbonSourcesController < ApplicationController
     def index
       if current_user.try(:admin?)
+        @carbon_sources = CarbonSource.all
         @carbon_source = CarbonSource.new
       else
         redirect_to root_path,
-          notice: "You are not authorized to access this page! Sorry'"
+          alert: "Sorry, you were not authorized to access that page!"
       end
     end
 
     def create
-      @carbon_source = CarbonSource.new(carbon_source_params)
-      if @carbon_source.save
+      @carbon_source = CarbonSource.new(new_carbon_source_params)
+      if current_user.try(:admin?) && @carbon_source.save
         redirect_to admin_carbon_sources_path,
           notice: "#{@carbon_source.source} added as carbon source"
+      elsif !current_user.try(:admin?)
+        redirect_to root_path,
+          alert: "Sorry, you were not authorized to access that action!"
       else
         @errors = @carbon_source.errors.full_messages
         render :index
       end
     end
 
+    def edit
+      if current_user.try(:admin?)
+        @carbon_source = CarbonSource.find(params[:id])
+      else
+        redirect_to root_path,
+          alert: "Sorry, you were not authorized to access that page!"
+      end
+    end
+
+    def update
+      @carbon_source = CarbonSource.find(params[:id])
+      if current_user.try(:admin?) &&
+        @carbon_source.update(edit_carbon_source_params)
```

Comment:
Use 4 (not 2) spaces for indenting a condition in an `if` statement spanning multiple lines.